pool-payout-zkapp/src/PoolPayout.test.ts

```diff
@@ -52,74 +76,489 @@ describe('pool payout', () => {
-    })
-
-    let signedData: Field[] = [];
-    rewardFields.rewards.forEach((reward) => {
-      signedData = signedData.concat(reward.index.toFields()).concat(reward.publicKey.toFields()).concat(reward.rewards.toFields())
-    })
-    signedData = signedData.concat(startingEpoch.toFields()).concat(feePayout.numDelegates.toFields()).concat(feePayout.payout.toFields().concat(validatorPublicKey.toFields()));
-
-    // TODO temp this is the same as in the lambda-payouts oracle but need to find a different way
-    const signature = Signature.create(
-      PrivateKey.fromBase58(ORACLE_PRIVATE_KEY_TESTING),
-      signedData
-    )
-
-    // Make the payouts
-    let tx2 = await Mina.transaction({ feePayerKey: deployerPrivateKey, fee: 1_000_000_000 }, () => {
-      pool.sendReward(rewardFields, feePayout, startingEpoch, startingIndex, validatorPublicKey, signature);
+    beforeEach(() => {
+      rewardFields = { rewards: [] };
+      for(let i=0; i<poolPayoutConfig.rewardsArrayLength; i++) {
+        rewardFields.rewards.push(Reward.blank());
+      }
+
+      feePayout = new FeePayout({
+        numDelegates: Field(0),
+        payout: UInt64.from(1000).mul(1000), // TODO while testing use 1000th of the rewards to make it easy
+      })
+
+      signedData = [];
+      rewardFields.rewards.forEach((reward) => {
+        signedData = signedData.concat(reward.index.toFields()).concat(reward.publicKey.toFields()).concat(reward.rewards.toFields())
+      })
+      signedData = signedData.concat(Field(poolPayoutConfig.deployEpoch).toFields()).concat(feePayout.numDelegates.toFields()).concat(feePayout.payout.toFields().concat(validatorPublicKey.toFields()));
+
+      if(poolPayoutConfig.oraclePrivateKey) {
+        signature = Signature.create(
+          PrivateKey.fromBase58(poolPayoutConfig.oraclePrivateKey),
+          signedData
+        )
+      } else {
+        throw("Cannot test without the oracle private key");
+      }
+    });
+
+    describe('invalid usage', () => {
+      it('throws for invalid epoch', async () => {
+        const invalidEpoch = Field(0);
+
+        await expect(async () => {
+          await Mina.transaction({ feePayerKey: deployerPrivateKey, fee: 1_000_000_000 }, () => {
+            pool.sendReward(rewardFields, feePayout, invalidEpoch, Field(poolPayoutConfig.deployIndex), validatorPublicKey, signature);
+          });
+        }).rejects.toThrow("The epoch must match");
+      });
+
+      it('throws for invalid index', async () => {
+        const invalidIndex = Field(999);
+
+        let tx = await Mina.transaction({ feePayerKey: deployerPrivateKey, fee: 1_000_000_000 }, () => {
+          pool.sendReward(rewardFields, feePayout, Field(poolPayoutConfig.deployEpoch), invalidIndex, validatorPublicKey, signature);
+        });
+        tx.sign([deployerPrivateKey]);
+        await tx.prove();
+
+        // Interesting that this throws later than the other errors... Might be good practice to catch this earlier
```

Comment:
This throws when the block is included and not via the contract, as it's just setting a precondition on the state. This is to allow multiple transactions in the same block.

Reply:
This is different from how the epoch handles it. That actually asserts against the on-chain state so it's a little inconsistent but this means that an epoch has to be completed (and included in a block) before the next one can be started. Normally this should be the expected behavior.